XX_BubbleSort/main.cpp

```diff
@@ -0,0 +1,449 @@
+					options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_NON_SEMANTIC_BIT;
+				// if you don't set the logger and source identifier you'll have no meaningful errors
+				options.preprocessorOptions.sourceIdentifier = "embedded.comp.hlsl";
+				options.preprocessorOptions.logger = m_logger.get();
+				options.preprocessorOptions.extraDefines = { &WorkgroupSizeDefine,&WorkgroupSizeDefine + 1 };
+				if (!(evenPassCPUShader = compiler->compileToSPIRV(evenPass, options)))
+					return logFail("Failed to compile following HLSL Shader:\n%s\n", evenPass);
+				if (!(oddPassCPUShader = compiler->compileToSPIRV(oddPass, options)))
+					return logFail("Failed to compile following HLSL Shader:\n%s\n", oddPass);
+			}
+
+			// Note how each ILogicalDevice method takes a smart-pointer r-value, so that the GPU objects refcount their dependencies
+			smart_refctd_ptr<nbl::video::IGPUShader> evenPassShader = device->createShader(evenPassCPUShader.get());
+			if (!evenPassShader)
+				return logFail("Failed to create a GPU Shader, seems the Driver doesn't like the SPIR-V we're feeding it!\n");
+
+			smart_refctd_ptr<nbl::video::IGPUShader> oddPassShader = device->createShader(oddPassCPUShader.get());
+			if (!oddPassShader)
+				return logFail("Failed to create a GPU Shader, seems the Driver doesn't like the SPIR-V we're feeding it!\n");
+
+			// the simplest example would have used push constants and BDA, but RenderDoc's debugging of that sucks, so I'll demonstrate "classical" binding of buffers with descriptors
+			nbl::video::IGPUDescriptorSetLayout::SBinding bindings[1] = {
+				{
+					.binding = 0,
+					.type = nbl::asset::IDescriptor::E_TYPE::ET_STORAGE_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE, // not is not the time for descriptor indexing
+					.stageFlags = IGPUShader::ESS_COMPUTE,
+					.count = 1,
+					.samplers = nullptr // irrelevant for a buffer
+				}
+			};
+			smart_refctd_ptr<IGPUDescriptorSetLayout> dsLayout = device->createDescriptorSetLayout(bindings);
+			if (!dsLayout)
+				return logFail("Failed to create a Descriptor Layout!\n");
+
+			const nbl::asset::SPushConstantRange pcRange = { .stageFlags = IShader::ESS_COMPUTE,.offset = 0,.size = sizeof(PushConstantData) };
+
+			// Nabla actually has facilities for SPIR-V Reflection and "guessing" pipeline layouts for a given SPIR-V which we'll cover in a different example
+			smart_refctd_ptr<nbl::video::IGPUPipelineLayout> pplnLayout = device->createPipelineLayout({ &pcRange, 1 }, smart_refctd_ptr(dsLayout));
+			if (!pplnLayout)
+				return logFail("Failed to create a Pipeline Layout!\n");
+
+			// We use strong typing on the pipelines (Compute, Graphics, Mesh, RT), since there's no reason to polymorphically switch between different pipelines
+			smart_refctd_ptr<nbl::video::IGPUComputePipeline> evenPipeline;
+			smart_refctd_ptr<nbl::video::IGPUComputePipeline> oddPipeline;
```

Comment:
Also you can use `pushConstants` to control odd vs. even, no need for separate pipelines.